docs: add indexer client example and docs (#1281)
latticexyz/mud: docs/pages/indexer.mdx

### Using the indexer with TypeScript

> You can see an example of an indexer client [here](https://github.com/latticexyz/mud/tree/230810-indexer-ts/examples/indexer-client).

Reviewer: ```suggestion
You can see [an example of an indexer client in the MUD repo](https://github.com/latticexyz/mud/tree/main/examples/indexer-client).
```

two changes here:
- updates the link to the main branch location once this is merged
- generally should avoid using "here" as link text, so updates to link to wrap the subject we're pointing at/providing more info abot

To execute it:

1. Download and build the example.

   ```sh copy
   git clone https://@github.com/latticexyz/mud.git

>    cd mud/examples/indexer_client

Reviewer: ```suggestion
   cd mud/examples/indexer-client
```

  address: '0x5FbDB2315678afecb367f032d93F642f64180aa3',
  tableId: '0x00000000000000000000000000000000436f756e746572000000000000000000',
  namespace: '',
  name: 'Counter',
  keySchema: {},
  valueSchema: { value: 'uint32' },
  lastUpdatedBlockNumber: 2042n,
  records: [ { key: {}, value: [Object] } ]
}
The actual value: 2
```

#### Detailed explanation

```ts
import { createIndexerClient } from "@latticexyz/store-sync/trpc-indexer";
const indexer = createIndexerClient({
  url: "http://localhost:3001",
});
```

Create an indexer client. The URL, `http://localhost/3001`, is the one for the indexer you run on the local computer.
If the indexer is elsewhere, modify the URL as appropriate.

```ts
const result = await indexer.findAll.query({
  chainId: 31337,
  address: "0x5FbDB2315678afecb367f032d93F642f64180aa3",
});
```

Connect to the indexer.
The `chainId` value is for a sanity check, to ensure we are connecting to the correct indexer.
The `address` parameter is the address of the `World` contract whose information we are reading.
The same indexer can read from multiple `World` objects, as long as they are on the same blockchain.

```ts

> const blockNum = result.blockNumber;
> console.log(`Block number: ${blockNum}`);

Reviewer: ```suggestion
console.log(`Block number: ${result.blockNumber}`);
```

### Using the indexer with TypeScript

1. Create these files in a previously empty directory:

   ```json filename="package.json" showLineNumbers copy
   {
     "name": "client",
     "version": "0.0.0",
     "private": true,
     "license": "MIT",
     "type": "module",
     "scripts": {

>        "indexerClient": "npx ts-node --esm index.ts"

Reviewer: no need for `npx` here if you have `ts-node` as a dep

also we tend to use `tsx` as it has slightly better ergonomics

```suggestion
       "start": "tsx index.ts"
```

       "module": "ESNext",
       "moduleResolution": "Node",
       "strict": true,
       "esModuleInterop": true,
       "noEmit": true,
       "skipLibCheck": true
     }
   }
   ```

   ```ts filename="index.ts" showLineNumbers=4 copy
   import { createIndexerClient } from "@latticexyz/store-sync/trpc-indexer";
   const indexer = createIndexerClient({
     url: "http://localhost:3001",
   });

   const result = await indexer.findAll.query({
     chainId: 31337,
     address: "0x5FbDB2315678afecb367f032d93F642f64180aa3",
   });

   const blockNum = result.blockNumber;
   console.log(`Block number: ${blockNum}`);

   console.log(`Tables: ${result.tables.map((t) => t.name)}`);

   const counterTable = result.tables.filter((t) => t.name == "Counter")[0];
   console.log("Information about Counter");
   console.log(counterTable);

   console.log(`The actual value: ${counterTable.records[0].value.value}`);
   ```

1. Install the necessary libraries.

   ```sh copy
   pnpm install
   ```

1. Run the client

   ```sh copy
   pnpm indexerClient
   ```

   If you are running [the minimal template](/tutorials/minimal), the expected output should be similar to:

   ```

>    qbzzt1@penguin:~/clnt2$ pnpm indexerClient
> 
>    > client@0.0.0 indexerClient /home/qbzzt1/clnt2

Reviewer 2: let's keep this clean
```suggestion
   $ pnpm indexerClient

   > client@0.0.0 indexerClient
```

The same indexer can read from multiple `World` objects, as long as they are on the same blockchain.

```ts
const blockNum = result.blockNumber;
console.log(`Block number: ${blockNum}`);
```

The `result` has two fields:

- `blockNumber`, the data is correct as of that block (it could change later).
- `tables`, the tables of the `World` we are reading.

```ts
console.log(`Tables: ${result.tables.map((t) => t.name)}`);
```

Use [`map`](https://www.tutorialspoint.com/typescript/typescript_array_map.htm) to get the names of the tables.

```ts
const counterTable = result.tables.filter((t) => t.name == "Counter")[0];
```

Use [`filter`](https://www.tutorialspoint.com/typescript/typescript_array_filter.htm) to get the `Counter` table, assuming there is at least one.

```ts
console.log("Information about Counter");
console.log(counterTable);
```

Log the information we have about the table.
It contains these fields:

| Field                  | Meaning                                                                                       |
| ---------------------- | --------------------------------------------------------------------------------------------- |

> | id                     | The table's ID, which encodes the `World` address, the name space, and the name of the table. |

Reviewer: this is just an internal ID for uniqueness in the context of SQL, doesn't correspond to anything on chain or in the protocol

| address                | The `World` address                                                                           |

> | tableId                | The ASCII encoded name of the namespace and then table name                                   |

Reviewer: this is actually the `bytes32` hex encoded table ID (which is a concat of `bytes16(namespace)` + `bytes16(name)`)
